### DISCUSSION TITLE
Getting a Prerender Error during deployment as well as a type error

### DISCUSSION THREAD

**Discussion by 12Kentos** [vercel/community#6930] [Help] [answered]
### Summary

Hey everyone,

I'm building a headless site that is using WordPress as the cms. Using wpgraphql as a plugin. I'm running into a pre-render error/type error, when I try and upload my code to Vercel.

### Steps to Reproduce

During the build production that is automatically done, it gives me the following error. 

```
}
TypeError: Cannot read properties of undefined (reading 'toString')
    at /vercel/path0/.next/server/chunks/948.js:2:4371
    at /vercel/path0/.next/server/chunks/948.js:2:15874
    at a.with (/vercel/path0/.next/server/chunks/948.js:1:8878)
    at c.with (/vercel/path0/.next/server/chunks/948.js:1:3078)
    at c.startActiveSpan (/vercel/path0/.next/server/chunks/948.js:1:18133)
    at a.startActiveSpan (/vercel/path0/.next/server/chunks/948.js:1:18667)
    at /vercel/path0/.next/server/chunks/948.js:2:15407
    at a.with (/vercel/path0/.next/server/chunks/948.js:1:8878)
    at c.with (/vercel/path0/.next/server/chunks/948.js:1:3078)
    at b.trace (/vercel/path0/.next/server/chunks/948.js:2:15361) {
  digest: '2509559086'
}
Error occurred prerendering page "/how-it-works". Read more: https://nextjs.org/docs/messages/prerender-error
TypeError: Cannot read properties of undefined (reading 'toString')
    at /vercel/path0/.next/server/chunks/948.js:2:4371
    at /vercel/path0/.next/server/chunks/948.js:2:15874
    at a.with (/vercel/path0/.next/server/chunks/948.js:1:8878)
    at c.with (/vercel/path0/.next/server/chunks/948.js:1:3078)
    at c.startActiveSpan (/vercel/path0/.next/server/chunks/948.js:1:18133)
    at a.startActiveSpan (/vercel/path0/.next/server/chunks/948.js:1:18667)
    at /vercel/path0/.next/server/chunks/948.js:2:15407
    at a.with (/vercel/path0/.next/server/chunks/948.js:1:8878)
    at c.with (/vercel/path0/.next/server/chunks/948.js:1:3078)
    at b.trace (/vercel/path0/.next/server/chunks/948.js:2:15361)
 ✓ Generating static pages (7/7)
> Export encountered errors on following paths:
    /(routes)/how-it-works/page: /how-it-works
Error: Command "npm run build" exited with 1
```

The interesting thing, is while it's being run locally it doesn't have any issues. This includes running npm run build on my local machine. It runs and executes without any errors. The above error only happens when I'm uploading the code to Vercel.

Here's the code that's causing the issue to happen.

```
import styles from "./page.module.scss";

async function getPage() {
  const query = `
  query GetPostByUri {
    pageBy(uri: "distance-education/how-it-works/") {
        title
        content
        heroImage {
          heroImage {
            node {
              sourceUrl
              altText
            }
          }
        }
      }
  }
      `;

  const res = await fetch(process.env.NEXT_PUBLIC_GRAPHQL_ENDPOINT, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    next: {
      revalidate: 5,
    },
    body: JSON.stringify({ query }),
  });

  const responseBody = await res.json();

  if (responseBody && responseBody.data && responseBody.data.pageBy) {
    return responseBody.data.pageBy;
  } else {
    throw new Error("Failed to fetch the page");
  }
}

export default async function HowItWorks() {
  const page = await getPage();

  return (
    <div className={styles.wrapper}>
      <h1 className={styles.title}>{page.title}</h1>
      {page.heroImage.heroImage ? (
        <img
          src={page.heroImage.heroImage.node.sourceUrl}
          className={styles.heroImg}
          alt={page.heroImage.heroImage.node.altText}
        />
      ) : (
        <></>
      )}

      <div
        dangerouslySetInnerHTML={{ __html: page.content }}
        className={styles.content}
      />
    </div>
  );
}
```

Current file path is app > (routes) > how-it-works > page.js

From the testing I've done, the code uploads to Vercel just fine when it's in a dynamic path.

For instance file path would be app > (routes) > testing > academy > [uri] > page.js

However when the I follow the path on the live site it runs into the same error.

The Vercel Deployment is connected to my Github account, so anytime I push changes there it automatically does a new deployment, and when it tries to build during that deployment this is when the error occurs.

I can't figure out what is causing this to happen. Any help would be greatly appreciated!

**Comment by versecafe** [comment]
Do you know where toString is being called in the code? or what function from a dep is calling it to help narrow the issue down?

**Reply by 12Kentos** [reply]
Unfortunately no, that's one of the reasons I've been unable to figure this out. Can't pinpoint where in the code/dep the error is coming from. I've also tried removing the return code, in the HowItWorks function and leaving it as an empty div to make sure nothing is going on there. So I believe it's somewhere in the `getPage()` function that is causing this issue.

I've also tried changing the fetch to a GET from the current POST method, and it gave me a different error. I've attached pictures of the error and code associated with the GET method.

![New error](https://github.com/vercel/community/assets/34455912/d26a09f2-07d8-443d-9616-df011ecd6793)
![get method](https://github.com/vercel/community/assets/34455912/0e175689-82d9-4851-ae68-b206598fd96d)

**Comment by versecafe** [comment]
can you try logging out step by step indicators to help narrow it down, also with the heroImage ? your end should return (null) instead of (<></>)

**Reply by 12Kentos** [reply]
Step by step indicators like?

`  console.log(query);`

Also I'm not sure it needs a null, as it's only part of the component that isn't being rendered if it's not there? (https://legacy.reactjs.org/docs/conditional-rendering.html#preventing-component-from-rendering)

Seems like it's not something you typically do?

**Reply by versecafe** [reply]
just like

```
some code

console.log("1");

more code

console.log("2");
```

just kind of helping to find at what point it breaks, also it *shouldn't* need to be null it's just a bit safer to be explicit

**Reply by 12Kentos** [reply]
That makes sense.

As for the indicators I've tried doing that and it doesn't really help with this situation as the error never shows up unless I'm trying to deploy the site to be hosted. The error only shows up when I've pushed my code to github, and Vercel (or Netlify) try and build the project to be deployed. That's when the error is showing up, it has never shown up when it's run locally on my machine. Everything works fine when running npm run dev, as well as npm run build. Neither of those run into any issues locally.

**Reply by versecafe** [reply]
with the [Vercel CLI](https://vercel.com/docs/cli) use `vercel dev` to try and emulate the Vercel deploy or `vercel` to do a test deployment on vercel so you don't need to make commits to test

**Reply by 12Kentos** [reply]
I've adjusted the code as follows.

```
import styles from "./page.module.scss";

async function getPage() {
  // pageBy(uri: "distance-education/how-it-works/") {
  const query = `
  query GetPostByUri {
    pageBy(uri: "third-kent-testing/") {
        title
        content
        heroImage {
          heroImage {
            node {
              sourceUrl
              altText
            }
          }
        }
      }
  }
      `;

  console.log("query below");
  console.log(query);

  const res = await fetch(process.env.NEXT_PUBLIC_GRAPHQL_ENDPOINT, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    next: {
      revalidate: 5,
    },
    body: JSON.stringify({ query }),
  });

  console.log("res below");
  console.log(res);

  const responseBody = await res.json();

  console.log("responseBody below");
  console.log(responseBody);

  if (responseBody && responseBody.data && responseBody.data.pageBy) {
    return responseBody.data.pageBy;
  } else {
    throw new Error("Failed to fetch the page");
  }
}

export default async function HowItWorks() {
  const page = await getPage();

  console.log("page below");
  console.log(page);

  return (
    <div className={styles.wrapper}>
      <h1 className={styles.title}>{page.title}</h1>
      {page.heroImage.heroImage ? (
        <img
          src={page.heroImage.heroImage.node.sourceUrl}
          className={styles.heroImg}
          alt={page.heroImage.heroImage.node.altText}
        />
      ) : (
        <></>
      )}

      <div
        dangerouslySetInnerHTML={{ __html: page.content }}
        className={styles.content}
      />
    </div>
  );
}
```

While uploading to vercel (Sorry haven't done the cli yet, I'll do that next)

I got the following error, right after the `console.log(query);`

This leads me to believe the fetch is having the issue. I can't figure out what is being marked as undefined in that section of code.

![Error with console logs](https://github.com/vercel/community/assets/34455912/120abc7f-c07f-45dc-823d-e805f758abe0)

**Reply by 12Kentos** [reply, marked as answer]
Vercel needed env variables added. Fixed the issue.... :/

**Reply by 12Kentos** [reply]
Thank you so much for your help!!!

**Reply by versecafe** [reply]
no problem, also the env var thing, we've all been there super easy mistake to make.
